katdal/datasources.py

```diff
@@ -53,24 +151,73 @@ def name(self):
 
 class TelstateDataSource(DataSource):
     """A data source based on :class:`katsdptelstate.TelescopeState`."""
-    def __init__(self, telstate, source_name='telstate'):
+    def __init__(self, telstate, capture_block_id=None, stream_name=None,
+                 source_name='telstate'):
+        # Create telstate view based on capture block ID and/or stream name
+        if stream_name:
+            telstate = telstate.view(stream_name)
+        if capture_block_id:
+            telstate = telstate.view(capture_block_id)
+        if stream_name and capture_block_id:
+            cb_stream = telstate.SEPARATOR.join((capture_block_id, stream_name))
+            telstate = telstate.view(cb_stream)
         self.telstate = telstate
-        attrs = {}
+        # Collect sensors
         sensors = {}
         for key in telstate.keys():
-            if telstate.is_immutable(key):
-                attrs[key] = telstate[key]
-            else:
+            if not telstate.is_immutable(key):
                 sensors[key] = TelstateSensorData(telstate, key)
-        metadata = AttrsSensors(attrs, sensors, name=source_name)
-        DataSource.__init__(self, metadata, None)
+        metadata = AttrsSensors(telstate, sensors, name=source_name)
+        try:
+            base_name = telstate['chunk_name']
+            chunk_info = telstate['chunk_info']
+        except KeyError:
+            # Metadata without data
+            DataSource.__init__(self, metadata, None)
+        else:
+            # Extract VisFlagsWeights and timestamps from telstate
+            store = RadosChunkStore.from_config(telstate['ceph_conf'],
+                                                telstate['ceph_pool'])
+            ts_name = store.join(base_name, 'timestamps')
+            ts_chunks = chunk_info['timestamps']['chunks']
+            ts_dtype = chunk_info['timestamps']['dtype']
+            timestamps = store.get_dask_array(ts_name, ts_chunks, ts_dtype)
+            timestamps = DaskLazyIndexer(timestamps)
+            data = ChunkStoreVisFlagsWeights(store, base_name, chunk_info)
+            DataSource.__init__(self, metadata, timestamps, data)
+
+    @classmethod
+    def from_url(cls, url):
+        """Construct TelstateDataSource from URL (RDB file / REDIS server)."""
+        url_parts = urlparse.urlparse(url, scheme='file')
+        kwargs = dict(urlparse.parse_qsl(url_parts.query))
+        kwargs['source_name'] = url_parts.geturl()
+        if url_parts.scheme == 'file':
+            # RDB dump file
+            telstate = katsdptelstate.TelescopeState()
+            try:
+                telstate.load_from_file(url_parts.path)
+            except OSError as err:
+                raise DataSourceNotFound(str(err))
+            return cls(telstate, **kwargs)
+        elif url_parts.scheme == 'redis':
+            # Redis server
+            try:
+                telstate = katsdptelstate.TelescopeState(url_parts.netloc)
```

Comment:
I know redis encodes the db number in the URL, I think in the query string. We should be able to pass that through.

Reply:
Have we ever used a db other than 0?

Reply:
Not yet, but since telstate supports it we may as well carry the support through rather than silently doing the wrong thing.

Reply:
I now extract db from the query and default to 0 if not there - is that OK? (One extra line of code vs a few more if I need to defer to TelescopeState's default db value).

Reply:
Sounds fine to me.